captum/attr/_core/multiscale_fast_cam.py

```diff
@@ -0,0 +1,400 @@
+                        the remaining dimensions will match the input.
+                        If a single tensor is provided as inputs, a single tensor is
+                        returned. If a tuple of tensors is provided for inputs, a
+                        tuple of corresponding sized tensors is returned.
+
+        Examples::
+
+            >>> # ImageClassifier takes a single input tensor of images Nx3x32x32,
+            >>> # and have multiple convolutional layers. Suppose we combine the
+            >>> # saliency map of three ReLU layers, which are
+            >>> # instances of nn.conv2d.
+            >>> # We will also use `smoe` scale and `gaussian` normalization,
+            >>> # which often yields the best results.
+            >>> net = ImageClassifier()
+            >>> fastcam = MultiscaleFastCam(net, [net.conv1, net.conv2, net.conv3])
+            >>> input = torch.randn(1, 3, 32, 32, requires_grad=True)
+            >>> # We set here that each layer we selected contributes equally to our
+            >>> # combined saliency map.
+            >>> attribution = fastcam.attribute(input, scale='smoe', norm='gaussian')
+            >>> # If we want to obtain each layer's attribution individually, we
+            >>> # set `combine=False`
+            >>> attribution = fastcam.attribute(input, scale='smoe',
+                                                norm='gaussian', combine=False)
+        """
+        # pick out functions
+        self.scale_func = self.pick_scale_func(scale)
+        self.norm_func = self.pick_norm_func(norm)
+        if not weights:
+            weights = np.ones(len(self.layers))
+
+        layer_attrs = self.layer_act.attribute(
+            inputs, additional_forward_args, attribute_to_layer_input
+        )
+        attributes = []
+        for layer_attr in layer_attrs:
+            scaled_attr = self.scale_func(layer_attr)
+            normed_attr = self.norm_func(scaled_attr)
+            attributes.append(normed_attr)
+        attributes = tuple(attributes)
+
+        # Combine
+        bn, channels, height, width = inputs.shape
+        combined_map = torch.zeros(
+            (bn, 1, height, width),
+            dtype=attributes[0].dtype,
+            device=attributes[0].device,
+        )
+        weighted_maps = [[] for _ in range(bn)]  # type: List[List[Any]]
+        for m, smap in enumerate(attributes):
+            for i in range(bn):
+                w = F.interpolate(
+                    smap[i].unsqueeze(0).unsqueeze(0),
+                    size=(height, width),
+                    mode=resize_mode,
+                    align_corners=False,
+                ).squeeze()
+                weighted_maps[i].append(w)
+                combined_map[i] += w * weights[m]
+        combined_map = combined_map / np.sum(weights)
+        weighted_maps = torch.stack([torch.stack(wmaps) for wmaps in weighted_maps])
```

Comment:
It looks like this will now be shaped as num_examples x num_layers x 1 x height x width, is that correct? We should at least document this in the return type, since this would have one extra dimension than the input. Alternatively, the extra dimension of one could be avoided, or we could return a list of num_layers with each layer separately analogous to the layer methods.